### DISCUSSION TITLE
Transforming Our Project into a Fullstack Application

### DISCUSSION THREAD

**Discussion by tjbck** [open-webui/open-webui#75] [Polls]
To implement the features mentioned in #68, #49, #31, we would have to turn this project into a fullstack application.

**Comment by SethBurkart123** [comment]
I'm definitely interested in the possibility of a RAG implementation, it would allow the webui to be significantly more useful in comparison to other publicly available chatbot implementations.

**Comment by coolaj86** [comment]
If you want to create something to manage users and access, keep that separate from the core WebUI.

It's easy enough to allow OIDC / OAuth2 session token passing to be put in a `Authorization: Bearer <token>` header (or to use Cookies, now that we have `Same-Site` and other mechanisms to pretty much eliminate the security risk they used to have).

Then you can have SSO on one page, User Management on another, and the Chat UI stays just that - a Chat UI.

You can surface payment needed and similar errors using OIDC  / OAuth2 spec `error` (code), `error_description` (what the user sees), and `error_uri` (where the user should be directed, i.e. for payment).

The value in this Ollama Chat UI project [compared to the others](https://github.com/jmorganca/ollama/issues/875#issuecomment-1778045758) is its simplicity - that it _can_ be built static and used immediately.

**Comment by coolaj86** [comment]
P.S. If you start to get buried down in the concerns of how business owners want you to change their product to build their user models around it, you might consider offering consulting.

And on that note. I would _love_ to have the option to pay for a commercial license for this to continue to be an embeddable tool - just no have a sense of security that it will be around in the future. I bet others would too.

Depends on what you want.
- If you want more stars, keep adding more and broader feature sets
- If you want a sustainable project, offer a commercial license and continue to make it the best Chat UI \
  (don't get distracted with non-unique value)

**Comment by TheRouce** [comment]
I'd say adding many ways to donate, consulting, or even just a call for more developers on social media like reddit is a great way to start. I'm a bit of an open source purist and would hate to see a commercial license attached to this, but honestly I wouldn't necessarily fault y'all if you want to see this become a main source of income, I just urge you to seek consulting first.

**Off topic:** I would love to help work on any future features, I am definitely not the strongest Svelte dev and would use this project to learn. But, if y'all want a website to show off the web ui then I'd be more than happy to maintain and host a website, possibly pay for a domain.

**Comment by kinthaiofficial** [comment]
Transforming Open WebUI into a proper fullstack application opens up some interesting architectural decisions. The key areas:

**State persistence beyond the session**
Moving conversation history, user preferences, and model configurations from in-memory/localStorage to a database backend. This enables: cross-device sync, conversation sharing, admin visibility into usage.

**User management and authentication**
Who can access which models? Rate limits per user? The auth layer becomes important when you have multiple users sharing a deployment. SSO integration (SAML, OIDC) for enterprise.

**Multi-model routing and load balancing**
In a multi-user deployment, requests to expensive models should be queued and rate-limited. A backend job queue prevents everyone hitting GPT-4 simultaneously.

**API-first design**
Make every UI capability accessible via API. This enables: programmatic conversation management, integration with external tools, agent systems that need to read/write conversations.

**Observability backend**
Log all model calls with: user, model, input tokens, output tokens, latency, cost. Expose aggregate dashboards. Without this, you can't bill users or understand which models are actually being used.

**Plugin/extension system**
The fullstack version could support server-side plugins (tool integrations, custom RAG pipelines, agent capabilities) that aren't possible in a purely client-side app.

**The minimum viable fullstack move:**
1. Add PostgreSQL/SQLite for persistent storage
2. Add a simple auth layer (email+password or OAuth)
3. Persist conversations to DB
4. Add an admin dashboard for usage monitoring

The rest can come incrementally. What's the specific capability that's blocked by the current architecture?
